File: go-backend-interview-prep/04_interfaces_and_generics/01_interfaces_basics/main.go
package main

import (
	"fmt"
	"math"
)

// --- Interface definition ---
type Shape interface {
	Area() float64
	Perimeter() float64
}

// --- Circle satisfies Shape implicitly ---
type Circle struct {
	Radius float64
}

func (c Circle) Area() float64 {
	return math.Pi * c.Radius * c.Radius
}

func (c Circle) Perimeter() float64 {
	return 2 * math.Pi * c.Radius
}

// --- Rectangle also satisfies Shape ---
type Rectangle struct {
	Width, Height float64
}

func (r Rectangle) Area() float64 {
	return r.Width * r.Height
}

func (r Rectangle) Perimeter() float64 {
	return 2 * (r.Width + r.Height)
}

// --- Function that accepts the interface ---
func printShape(s Shape) {
	fmt.Printf("  area=%.2f  perimeter=%.2f\n", s.Area(), s.Perimeter())
}

func main() {
	// --- Implicit satisfaction ---
	fmt.Println("--- Interface basics ---")
	c := Circle{Radius: 5}
	r := Rectangle{Width: 10, Height: 3}

	printShape(c) // Circle satisfies Shape
	printShape(r) // Rectangle satisfies Shape

	// --- Interface variable ---
	fmt.Println("\n--- Interface variable ---")
	var s Shape
	s = c
	fmt.Printf("type holding Circle: area=%.2f\n", s.Area())
	s = r
	fmt.Printf("type holding Rect:   area=%.2f\n", s.Area())

	// --- Slice of interfaces ---
	fmt.Println("\n--- Slice of interfaces ---")
	shapes := []Shape{
		Circle{Radius: 1},
		Rectangle{Width: 4, Height: 5},
		Circle{Radius: 3},
	}
	for _, sh := range shapes {
		printShape(sh)
	}

	// --- Nil interface trap ---
	fmt.Println("\n--- Nil interface trap ---")
	var nilShape Shape
	fmt.Println("nil interface:", nilShape == nil) // true

	var cp *Circle // nil pointer
	nilShape = cp  // storing nil pointer in interface
	fmt.Println("nil pointer in interface:", nilShape == nil) // FALSE!
	// The interface holds (*Circle, nil) -- it is non-nil
}

File: go-backend-interview-prep/04_interfaces_and_generics/01_interfaces_basics/main.py
# Python equivalent of Interfaces Basics (compare with main.go)
# Go interfaces (implicit) ~ Python Protocol (structural typing, 3.8+)
# Python also has ABC, but Protocol is closer to Go's model.

import math
from typing import Protocol


# --- Protocol = structural interface (like Go interface) ---
class Shape(Protocol):
    def area(self) -> float: ...
    def perimeter(self) -> float: ...


# --- Circle satisfies Shape (no explicit declaration) ---
class Circle:
    def __init__(self, radius: float):
        self.radius = radius

    def area(self) -> float:
        return math.pi * self.radius ** 2

    def perimeter(self) -> float:
        return 2 * math.pi * self.radius


# --- Rectangle also satisfies Shape ---
class Rectangle:
    def __init__(self, width: float, height: float):
        self.width = width
        self.height = height

    def area(self) -> float:
        return self.width * self.height

    def perimeter(self) -> float:
        return 2 * (self.width + self.height)


def print_shape(s: Shape) -> None:
    """Accepts anything with area() and perimeter() methods."""
    print(f"  area={s.area():.2f}  perimeter={s.perimeter():.2f}")


def main():
    # --- Implicit satisfaction (duck typing / Protocol) ---
    print("--- Interface basics ---")
    c = Circle(radius=5)
    r = Rectangle(width=10, height=3)

    print_shape(c)
    print_shape(r)

    # --- Variable holding different types ---
    print("\n--- Variable holding different types ---")
    s: Shape = c
    print(f"type holding Circle: area={s.area():.2f}")
    s = r
    print(f"type holding Rect:   area={s.area():.2f}")

    # --- List of shapes ---
    print("\n--- List of shapes ---")
    shapes: list[Shape] = [
        Circle(radius=1),
        Rectangle(width=4, height=5),
        Circle(radius=3),
    ]
    for sh in shapes:
        print_shape(sh)

    # --- None vs nil interface ---
    print("\n--- None comparison ---")
    nil_shape = None
    print("None shape:", nil_shape is None)  # True
    # Python has no nil-pointer-in-interface trap.
    # None is just None. No type-value pair wrapper.


if __name__ == "__main__":
    main()
